### PR TITLE
Adding Efficiency and Delta R plots

### PR DESCRIPTION
**Pull request by samf25**
I needed Efficiency and Delta R plots for some comparisons I was working on. Adding these to main will allow anyone to be able to use them.

### CODE REVIEW

File: src/TruthHists.cxx [hunk 1]
```diff
@@ -32,3 +43,33 @@ void TruthHists::fill(const EVENT::MCParticle* particle) {
...
+void TruthHists::deltaR(const EVENT::MCParticle* particle, const EVENT::Track* track) {
+	const double* mom = particle->getMomentum();
+        double pt = std::sqrt(std::pow(mom[0], 2) + std::pow(mom[1], 2));
+        double eta = std::atanh(mom[2] / std::sqrt(std::pow(pt, 2) + std::pow(mom[2], 2)));
+        double trackEta = -std::log(std::tan((1.57079632679 - std::atan(track->getTanLambda()))/2));
+        double phi = std::atan(mom[1] / mom[0]);
+        double trackPhi = track->getPhi();
+	double dphi = phi - trackPhi;
+	double deta = eta - trackEta;
+	if (dphi > 3) { dphi -= 3.1415926535; }
```
Comment (kkrizka) [inline src/TruthHists.cxx]: Use [TVector2::Phi_mpi_pi](https://root.cern.ch/doc/master/classTVector2.html#ad30d7548d7a2562d86cd8d1597460a44) instead.

File: TrackPerf/TruthHists.hxx [hunk 2]
```diff
@@ -14,10 +16,12 @@ class TruthHists {
...
 
   //! Initialize empty histograms
-  TruthHists();
+  TruthHists(bool effi);
 
   // Fill histograms with a single track
-  void fill(const EVENT::MCParticle* track);
+  void fill(const EVENT::MCParticle* particle);
+  void effi(const EVENT::MCParticle* particle, bool passed);
+  void deltaR(const EVENT::MCParticle* particle, const EVENT::Track* track);
```
Comment (kkrizka) [inline TrackPerf/TruthHists.hxx]: Histograms correlating `MCParticle` and `Track` properties should go into `ResoHists`.

File: src/TrackHists.cxx [hunk 3]
```diff
@@ -56,3 +62,11 @@ void TrackHists::fill(const EVENT::Track* track) {
     h_nhit3->Fill(subdetectorHitNumbers[5] + subdetectorHitNumbers[6]);
   }
 }
+
+void TrackHists::effi(const EVENT::Track* track, bool passed) {
+	float pt = fabs(0.3 * _Bz / track->getOmega() / 1000);
+	float eta = -std::log(std::tan((1.57079632679 - std::atan(track->getTanLambda()))/2));
```
Comment (kkrizka) [inline src/TrackHists.cxx]: Use [TMath::Piover2](https://root.cern.ch/doc/master/namespaceTMath.html#a705b6f0bb7c7aa1ab6f2c2e4aa2acc71).

File: src/TrackHists.cxx [hunk 4]
```diff
@@ -30,6 +30,12 @@ TrackHists::TrackHists() {
   h_nhit3 =
       new TH1F("reco_nhit_outer", ";Outer tracker track hits; Tracks [/hit]",
                20, -0.5, 19.5);
+  if (effi) {
+    h_effpt = new TEfficiency("fake_pt_vs_eff", 
+      "Fake Track Rate vs pT;Reconstructed pT [GeV]; Efficiency", 50, 0, 12);
+    h_effeta = new TEfficiency("fake_eta_vs_eff", 
+      "Fake Track Rate vs Eta;Reconstructed eta; Efficiency", 50, -4, 4);
```
Comment (kkrizka) [inline src/TrackHists.cxx]: Make plots vs theta instead to align it as all other plots currently use that unit.

File: src/TrackPerfHistProc.cxx [hunk 5]
```diff
@@ -54,23 +54,24 @@ void TrackPerfHistProc::init() {
...
+  _realTracks = std::make_shared<TrackPerf::TrackHists>(false);
+  _realTruths = std::make_shared<TrackPerf::TruthHists>(true);
   _realReso = std::make_shared<TrackPerf::ResoHists>();
   tree->mkdir("../fake");
   tree->cd("../fake");
-  _fakeTracks = std::make_shared<TrackPerf::TrackHists>();
+  _fakeTracks = std::make_shared<TrackPerf::TrackHists>(true);
   h_number_of_fakes = new TH1F("number_of_fakes",
                                ";Number of fake tracks;Events", 100, 0, 300000);
+  h_count = new TH1F("Count","C;C",20,0,10);
```
Comment (kkrizka) [inline src/TrackPerfHistProc.cxx]: What's this for?

**Comment by kkrizka** [comment]
The way that `TrackPerf` is currently designed is to write the histograms and then use a plotting package to make the necessary efficiency plots (ie: [TrackPerfPlots](https://github.com/kkrizka/TrackPerfPlots)). I am not against saving a `TEfficiency` directly, as it is indeed convenient. But it is a bit awkward to fit it into the current output structure (`real`, `fake`, `all`) which targets a specific matching of MCParticle and Track. This you can see via the need to have the `effi` argument to the histogram classes.

Maybe create a new class called `EfficiencyHists`?

File: src/TrackPerfHistProc.cxx (hunk 5 above)
Comment (samf25) [reply src/TrackPerfHistProc.cxx]: Ah sorry about that, this was from debugging an issue in the Gaudi version. Removed now!

**Comment by samf25** [comment]
Ok I've moved the efficiency plots to a new class that adds a new directory in the TFile and fixed the other issues you pointed out. Thanks for the help!

File: src/EfficiencyHists.cxx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+using namespace TrackPerf;
+
+EfficiencyHists::EfficiencyHists(bool effi) {
+  if (effi) {
+     h_effpt = new TEfficiency("pt_vs_eff", ";Truth pT [GeV]; Efficiency", 50, 0, 12 );
+     h_efftheta = new TEfficiency("theta_vs_eff", ";Truth theta; Efficiency", 50, -4, 4);
+  } else {
+    h_effpt = new TEfficiency("fake_pt_vs_eff",
+      "Fake Track Rate vs pT;Reconstructed pT [GeV]; Efficiency", 50, 0, 12);
```
Comment (kkrizka) [inline src/EfficiencyHists.cxx]: I think you also want to change the y-axis label from "Efficiency" to "Fake Rate".

File: src/EfficiencyHists.cxx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  if (fabs(theta) < 2*std::atan(std::exp(-2))) h_effpt->Fill(passed, pt);
+  if (pt > 0.5) h_efftheta->Fill(passed, theta);
+}
+
+void EfficiencyHists::fillMC(const EVENT::MCParticle* particle, bool passed) {
+  const double* mom = particle->getMomentum();
+  double pt = std::sqrt(std::pow(mom[0], 2) + std::pow(mom[1], 2));
+  double theta = std::atanh(mom[2] / pt);
+  if (pt > 0.5) h_efftheta->Fill(passed, theta);
```
Comment (kkrizka) [inline src/EfficiencyHists.cxx]: Remove hardcoded checks on the MC particle reconstructibility to make this consistent with the other plots saved in the processor.

If you are concerned about low efficiency in theta being related to low efficiency in pT, maybe also add a 2D TEfficiency plot?

Comment (samf25) [reply src/EfficiencyHists.cxx]: I guess that makes sense. I've always added pT cuts to position efficiency plots and vice versa in the past, but doesn't hurt anything to remove them. For right not I don't think I need the 2D plot, but if I do, I'll add it later.

**Pull request merged**